### DISCUSSION TITLE
So close to packing a piscina server app

### DISCUSSION THREAD

**Discussion by johnmanko** [webpack/webpack#16240] [Q&A]
One of the tricks I've learned with webpacking server apps is to keep a few packages as externals, such as `sharp` and `piscina`.  When introducing `piscina`, the worker path becomes a big issue.  Here is my usage of `piscina` and a `worker.ts`:

```
import { Service } from '@hapipal/schmervice';
import Piscina from 'piscina';
import { WorkerData, WorkerDataResult } from '../util/thread-workers/worker.model';

export class WorkerPoolService extends Service {

    private pool: Piscina;

    initialize() {
        // This import is not needed
        import('../util/thread-workers/worker');
        this.pool = new Piscina({
            filename: new URL("../util/thread-workers/worker", import.meta.url).href
        });
    }

    async run(data: WorkerData): Promise<WorkerDataResult> {
        return this.pool.run(data);
    }

    teardown() {
        return this.pool.destroy();
    }

}
```

Adding `import('../util/thread-workers/worker');` doesn't help.  The Typescript-compiled file retains it's `.ts` extension.

Dockerfile:
```
# syntax=docker/dockerfile:1

FROM node:18.3.0-alpine AS builder
WORKDIR /src

COPY ./server/package.json ./server/
RUN cd ./server && \
    npm i --global-style=true --no-safe  --force sharp && \
    npm i --global-style=true --no-safe  --force piscina && \
    npm install --force
COPY ./server/ ./server/
RUN npm run webpack-build --prefix ./server && \
    mkdir ./server/dist/node_modules && \
    cp -r ./server/node_modules/sharp ./server/dist/node_modules/ && \
    cp -r ./server/node_modules/piscina ./server/dist/node_modules/

FROM node:18.3.0-alpine
EXPOSE 5000
WORKDIR /app
COPY --chown=1000:1000 --from=builder /src/server/dist/ ./
USER 1000
CMD ["node", "./index.bundle.mjs"]  
```

Output worker and index bundle:
```
/app $ ls
e3509471f8692264408d.ts       index.bundle.mjs              node_modules
```

Grep worker in index:
```
/app $ grep -r -E -o ".{0,10}e3509471f8692264408d.{0,10}" index.bundle.mjs
orts=n.p+"e3509471f8692264408d.ts"},6778
```

Running application
```
/app $ node index.bundle.mjs
node:internal/event_target:908
  process.nextTick(() => { throw err; });
                           ^
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /app/e3509471f8692264408d.ts
``` 

webpack.config.mjs:
```
import path from 'path';
import webpack from 'webpack';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export default {
    
    target: 'node',
    entry: {
        index: './src/index.ts'
    },
    devtool: 'inline-source-map',
    node: {
        global: true,
        __filename: true,
        __dirname: true,
    },
    module: {
        parser: {
          javascript: {
            importMeta: false,
          },
        },
        rules: [
            {
                test: /\.html$/i,
                loader: "html-loader",
            },
            {
                test: /\.[jt]sx?$/,
                resolve: {
                    fullySpecified: false,
                },
            },
            {
                test: /\.tsx?$/,
                use: 'ts-loader',
                exclude: /node_modules/,
            },
            { test: /\.node$/, use: "node-loader" }
        ],
    },
    experiments: {
        topLevelAwait: true,
        outputModule: true // false for commmonjs
    },
    resolve: {
        extensions: ['.tsx', '.ts', '.js'],
    },
    output: {
        filename: '[name].bundle.mjs',
        path: path.resolve(__dirname, 'dist'),
        sourceMapFilename: '[name].[contenthash].bundle.map',
        chunkFormat: 'module',
        chunkFilename: '[name].[contenthash].mjs'
    },
    plugins: [
        new webpack.IgnorePlugin({
            resourceRegExp: /canvas/,
            contextRegExp: /jsdom$/
        }),
        new webpack.ContextReplacementPlugin(/require_optional/)
    ],
    externals: {
        'sharp': 'module sharp',
        'piscina': 'module piscina'
    }
};
```

The worker file `../util/thread-workers/worker.ts`/`/app/e3509471f8692264408d.ts` is indeed transformed from Typescript to javascript, so Typescript is hitting it.  The file extension is not updated.  In addition, I believe it's extension needs to be `.mjs`.

Of course, I could just "rename" the files, like such:

```
for file in *.ts; do 
    mv -- "$file" "${file%.ts}.mjs"
    sed -i "s/$file/${file%.ts}.mjs/" index.bundle.mjs
done
```

But that doesn't take care of the imports, which are not found because they've been bundled.  It's a mess.
Get this working would reduce my Docker image from 480MB to 230MB, so I definitely want to get this fixed.

**Comment by alexander-akait** [comment]
Do you looking for this https://webpack.js.org/configuration/resolve/#resolveextensionalias?

**Reply by johnmanko** [reply]
Could solve part of the issue, but still leaves resolving it's imports that have already been bundled.

**Reply by alexander-akait** [reply]
Can you create small reproducible test repo (github) and I will look at this

**Reply by johnmanko** [reply]
@alexander-akait 

This should cover it.  

```
npm run webpack-build
```

or build the Docker image.

https://github.com/johnmanko/webpacked-server

**Reply by alexander-akait** [reply]
When you have `new URL("./something", import.meta.url)`, webpack assumes it is `asset modules` and copy file (in your case also run `ts-loader`), that is why you have this.

So you can do it fully dynamically (+ bundle worker using multi compiler mode with `target: "webworker"`):
```js
let url = "../util/thread-workers/worker.js";

this.pool = new Piscina({
  filename: new URL(url, import.meta.url).href
});
```

Another solutions:
```js
{
  test: /worker\.ts$/,
  type: 'asset/resource',
  use: 'ts-loader',
  generator: {
    filename: '[hash].mjs[query]'
  }
},
```

But in this case you don't bundle worker at all, you just strip types and copy file, so `'../calc-util'` is unavalaible.

Can you say what you expected? If you want to bundle, you need to go with the first way (because you can't use `new Worker(...)`).

**Reply by johnmanko** [reply]
Going the multicompile route does generate 2 bundles.  Of course, index is looking for the wrong file:

```
> ls
index.bundle.mjs        worker.bundle.mjs
> grep -r -E -o ".{0,50}worker.{0,50}" index.bundle.mjs
index.bundle.mjs:l=new r.default({filename:new URL("../util/thread-workers/worker",import.meta.url).href})}async run(e){ret
``` 

**webpack.multi.cjs**:
```
module.exports = [
    require('./webpack.config.cjs'),
    require('./webpack.worker.config.cjs')
  ]
```

**webpack.worker.config.cjs**:
```
const path = require('path');
const webpack = require('webpack');

module.exports = {
    
    target: 'webworker',
    entry: {
        worker: './src/lib/util/thread-workers/worker.ts'
    },
    output: {
        filename: '[name].bundle.mjs',
        path: path.resolve(__dirname, 'dist'),
        sourceMapFilename: '[name].[contenthash].bundle.map',
        chunkFormat: 'module',
        chunkFilename: '[name].[contenthash].mjs'
    },
```

**Reply by johnmanko** [reply]
@alexander-akait 

My goal, here, is to test with non-packed tsc builds, yet still have the webpack/docker build work.  The worker url string seems to be the last of the hurdles.  I can't change the non-webpacked path since that's it's real path during testing, but I need to change it for the generated bundle.

**Reply by alexander-akait** [reply]
@johnmanko Please use https://webpack.js.org/configuration/other-options/#dependencies, i.e. bundle `worker` firstly, then bundle the main application (use the **bunded, not source** worker file in `new URL("../util/thread-workers/worker",import.meta.url)`).

> My goal, here, is to test with non-packed tsc builds, yet still have the webpack/docker build work. The worker url string seems to be the last of the hurdles. I can't change the non-webpacked path since that's it's real path during testing, but I need to change it for the generated bundle.

Do you want to test the bundled application (and worker)? I don't think you should change URL here

**Reply by johnmanko** [reply]
I want to test using the unbundled app, but deploy the bundled file.  This is where the worker file path/name becomes a conflict between the two (unbundled vs bundled).

**Reply by johnmanko** [reply]
Does webpack have any buildtime replacement mechanism?  Maybe file swapping?


```
        let url = import("./some-swapped-file");
        this.pool = new Piscina({
            filename: new URL(url, import.meta.url).href
        });
```

During development/source execution of non-bundled tsc output, `./some-swapped-file`:

```
export default "../util/thread-workers/worker"
```

During webpack build, replace `./some-swapped-file` file with:

```
export default "./worker.mjs"
```

**Reply by johnmanko** [reply]
Giving this a try:  https://webpack.js.org/plugins/normal-module-replacement-plugin/

**Reply by alexander-akait** [reply]
@johnmanko hm, webpack output is valid here (accroding your options - you need an extension and you bundle ESM), because you need URL here, sorry I can't undestand the problem - What URL you expected and where?

**Reply by johnmanko** [reply]
@alexander-akait I resolved the file reference issue using file swapping:

The server webpack config:
```
    plugins: [
        new webpack.NormalModuleReplacementPlugin(
          /(.*)\/worker-path\.config/,
          '../util/thread-workers/worker-path.bundle.config'
        ),
```

File that uses import path string (service):
```
import * as Schmervice from '@hapipal/schmervice';
import Piscina from 'piscina';
import { WorkerData, WorkerDataResult } from '../util/thread-workers/worker.model';
import WORKER_PATH from '../util/thread-workers/worker-path.config';  // <---- this

export class WorkerPoolService extends Schmervice.Service {

    private pool: Piscina;

    initialize() {
        let url: string = `${WORKER_PATH}`;
        this.pool = new Piscina({
            filename: new URL(url, import.meta.url).href
        });
    }

    async run(data: WorkerData): Promise<WorkerDataResult> {
        return this.pool.run(data);
    }

    teardown() {
        return this.pool.destroy();
    }

}
```

Non-bundled path `worker-path.config.ts`:
```
export default "../util/thread-workers/worker";
```

Bundled path `worker-path.bundle.config.ts`:
```
export default "./worker.bundle.mjs";
```

So, I can effectively reference the correct file regardless of loading from a bundle or source.

Unfortunately, the `worker.bundle` that's generated is not valid.  It's not even close to a bundle of the worker:

```
import{createRequire as e}from"module";import*as r from"sharp";var o={d:(e,r)=>{for(var t in r)o.o(r,t)&&!o.o(e,t)&&Object.defineProperty(e,t,{enumerable:!0,get:r[t]})},o:(e,r)=>Object.prototype.hasOwnProperty.call(e,r)};e(import.meta.url)("node:fs"),o.d({},{});
```

That's it.  When decoding the source map:

```
{ 
    "sources":[
        "webpack://server/webpack/bootstrap",
        "webpack://server/webpack/runtime/define property getters",
        "webpack://server/webpack/runtime/hasOwnProperty shorthand",
        "webpack://server/external node-commonjs \"node:fs\"",
        "webpack://server/external module \"sharp\""
    ],
    "sourcesContent":[
        "
        // The require scope
        var __webpack_require__ = {};
        ",
        "
        // define getter functions for harmony exports\n
        __webpack_require__.d = (exports, definition) => {
            for(var key in definition) {
                if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
                    Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
                }
            }
        };",
        "
        __webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
        ",
        "
        const __WEBPACK_NAMESPACE_OBJECT__ = __WEBPACK_EXTERNAL_createRequire(import.meta.url)(\"node:fs\");
        ",
        "
        var x = y => { var x = {}; __webpack_require__.d(x, y); return x; }
        var y = x => () => x
        const __WEBPACK_NAMESPACE_OBJECT__ = x({  });
        "
    ],
    "names":[
        "__webpack_require__",
        "exports",
        "definition",
        "key",
        "o",
        "Object",
        "defineProperty",
        "enumerable",
        "get",
        "obj",
        "prop",
        "prototype",
        "hasOwnProperty",
        "call",
        "__WEBPACK_EXTERNAL_createRequire",
        "import",
        "meta",
        "url",
        "d"
    ],
    "sourceRoot":""
}
```

**Reply by alexander-akait** [reply]
~~It means your file was handled as an asset module, not as JS module and was not bundled~~ I am wrong

Can you update your reproducible example with the example above and ping me, I will look

**Reply by johnmanko** [reply]
@alexander-akait You can check out the updated repo.  Build is same `npm run webpack-build`.  You'll notice that the generated `worker.bundle.mjs` is empty.  `npm run build` will just compile the Typescript.

**Reply by johnmanko** [reply]
@alexander-akait Just wondering if anything stood out in either my configuration or coding that would cause webpack to output an empty `worker.bundle.mjs` file?  I couldn't pinpoint anything.

**Reply by alexander-akait** [reply]
Sorry for delay, I will look at this in near future - today (wasn't able to be at the computer)

**Reply by alexander-akait** [reply]
@johnmanko Sorry for long delay :disappointed: Do you need help with something? I tried to run your repo and looks it works

**Reply by alexander-akait** [reply]
[webpacked-server.zip](https://github.com/webpack/webpack/files/9770911/webpacked-server.zip)

Please look at my solution, you can run `npm run start` (typescript without bundling) and `npm run webpack-start` (using webpack)
